lib/exercises.rb

```diff
@@ -6,17 +6,88 @@
 # Space Complexity: ?
 
 def grouped_anagrams(strings)
-  raise NotImplementedError, "Method hasn't been implemented yet!"
+  return [] if strings.length == 0
+
+  sorted_strings = strings.map { |string|
+    string.chars.sort.join
+  }
+
+  sorted_strings_hash = {}
+
+  sorted_strings.each { |string|
+    sorted_strings_hash[string] = []
+  }
+
+  sorted_strings.each_with_index { |string, index|
+    sorted_strings_hash[string] << strings[index]
+  }
+
+  final_array = []
+
+  sorted_strings_hash.each_value do |value|
+    final_array.push(value)
+  end 
+
+  return final_array
 end
 
-# This method will return the k most common elements
-# in the case of a tie it will select the first occuring element.
 # Time Complexity: ?
 # Space Complexity: ?
 def top_k_frequent_elements(list, k)
-  raise NotImplementedError, "Method hasn't been implemented yet!"
+  return [] if list.length == 0
+  frequency_hash = {}
+
+  list.each do |element|
+    if frequency_hash[element].nil?
+      frequency_hash[element] = 1
+    else
+      frequency_hash[element] += 1
+    end 
+  end 
+
+
+  frequency_hash_values = frequency_hash.values
+  
+  sorted_frequency_hash_values = frequency_hash_values.sort
+
+  
+  sliced_sorted_frequency_hash_values = ''
+  # if sorted_frequency_hash_values.first == sorted_frequency_hash_values.last && k == 1
+  #   return list.first
+  if sorted_frequency_hash_values.first == sorted_frequency_hash_values.last 
+    return frequency_hash.keys
+  end
+
+  
+  if k <= list.length
+    
+    sliced_sorted_frequency_hash_values = sorted_frequency_hash_values.slice(-(k), k)
+  end 
+
+
+  official_sliced_sorted_frequency_hash_values = sliced_sorted_frequency_hash_values.map do |value|
+    frequency_hash.key(value)
```

Comment:
This searches the values O(n) for a matching value and returns the key.  Since you're doing this k times, this is O(nk)